### DISCUSSION TITLE
How to solve this in Prisma, count related entries of complex query

### DISCUSSION THREAD

**Discussion by riccardolardi** [prisma/prisma#25782] [Q&A] [answered]
### Question

Consider a filtering mechanism as suggested in the linked image. We would like to show the results count depending on each "attribute option". See the prisma schema and trpc route here: https://gist.github.com/riccardolardi/7ab7c9a7792ed9428298630c3b30f22c

This approach is evidently very inefficient since it introduces a classic n+1 problem, where we fetch the attributes for the selected category and then loop over each of them and again fetch the count for each attribute option.

I currently can't think of any other solution other than writing a raw SQL query for this kind of problem. Or am I missing something?

<img width="899" alt="Screenshot 2024-12-03 at 10 05 51" src="https://github.com/user-attachments/assets/65c83d23-fd3e-4ae2-9635-ad4d20913cbd">

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: macOS
- Database: PostgreSQL
- Node.js version: 20.17
- Prisma 5

**Comment by ludralph** [comment, marked as answer]
Hi @riccardolardi 

The filtered relation count feature can help to some extent, but it may not completely solve your specific use case in a single query. Here's why:

- Filtered Relation Count: This feature allows you to count related records with conditions, as described in the [documentation](https://www.prisma.io/docs/orm/prisma-client/queries/aggregation-grouping-summarizing#filter-the-relation-count). However, this feature doesn't fully solve your complex scenario where you need counts for each attribute option across multiple attributes.

One workaround is to reduce the number of queries by fetching all relevant posts with their attribute values in a single query, and then performing the counting logic in your application code. This would be more efficient than the current N+1 approach. You may also consider using raw SQL as that would be the most performant for you.

**Reply by riccardolardi** [reply]
Thanks a lot for the thorough reply

**Reply by ludralph** [reply]
You're welcome.
